In [65]:
from csv import DictReader
from numpy import *
from statistics import pvariance

In [94]:
A = []
B = []

In [95]:
with open('2.2.csv') as csv_file:
    csv_reader = csv.DictReader(csv_file)
    for row in csv_reader:
        A.append(float(row['A']))
        B.append(float(row['B']))
print(A)
print(B)

[3.0, 4.0, 6.0, 7.0, 9.0, 10.0, 9.0, 11.0, 12.0, 13.0, 15.0, 4.0]
[8.0, 9.0, 10.0, 13.0, 15.0, 14.0, 13.0, 16.0, 13.0, 19.0, 6.0, 19.0]


In [96]:
print(pvariance(A))
print(pvariance(B))

13.57638888888889
15.409722222222221


La matrice de variance-covariance a les variances pour diagonale :

In [97]:
V = cov(A, B, bias=True)
print(V)
covAB = V[0][1]
varA = V[0][0]
varB = V[1][1]

[[13.57638889  1.46527778]
 [ 1.46527778 15.40972222]]


In [98]:
EA = mean(A)
EB = mean(B)
stdDevA = sqrt(varA)
stdDevB = sqrt(varB)
corr = covAB / (stdDevA * stdDevB)
print(covAB)
print(corr)

1.4652777777777772
0.10130489326551938


La corrélation entre A et B est très faible, le modèle linéaire est donc inadapté.

In [99]:
B2approx = covAB/varA
B1approx = EB - B2approx * EA
print(B1approx)
print(B2approx)

11.990281329923274
0.10792838874680304


On obtient la droite de régression y = 11.99 + 0.11 x (qui ne nous servira pas à grand-chose, vu la qualité de l’approximation)

Effaçons maintenant les deux dernières observations.

In [100]:
del A[10:12]
del B[10:12]
print(A)
print(B)

[3.0, 4.0, 6.0, 7.0, 9.0, 10.0, 9.0, 11.0, 12.0, 13.0]
[8.0, 9.0, 10.0, 13.0, 15.0, 14.0, 13.0, 16.0, 13.0, 19.0]


Et c’est reparti.

In [103]:
V = cov(A, B, bias=True)
covAB = V[0][1]
varA = V[0][0]
varB = V[1][1]
EA = mean(A)
EB = mean(B)
stdDevA = sqrt(varA)
stdDevB = sqrt(varB)
corr = covAB / (stdDevA * stdDevB)
print(corr)

0.898205382062774


Ah-hah, la corrélation est nettement meilleure quand on enlève les points dérangeants.

In [104]:
B2approx = covAB/varA
B1approx = EB - B2approx * EA
print(B1approx)
print(B2approx)

5.470119521912349
0.8964143426294822


On obtient la droite de régression y = 5.47 + 0.90 x (mais sa validité repose sur la raison d’enlever les deux dernières données)